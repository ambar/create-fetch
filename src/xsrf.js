import toHeaders from './utils/toHeaders'
import toObject from './utils/toObject'

const getCookie = key => {
  const [, value] = document.cookie.match(new RegExp(`${key}=([^;]+)`)) || []
  return value
}

const xsrfMethods = ['DELETE', 'PATCH', 'POST', 'PUT']

const xsrf = ({
  cookieName = '_xsrf',
  headerName = 'x-xsrftoken',
} = {}) => fetch => (url, options = {}) => {
  const headers = toHeaders(options.headers)
  if (
    typeof document !== 'undefined' &&
    options.method &&
    xsrfMethods.indexOf(options.method.toUpperCase()) !== -1
  ) {
    const xsrfToken = getCookie(cookieName)
    if (xsrfToken) {
      headers.set(headerName, xsrfToken)
    }
  }

  return fetch(url, {
    ...options,
    headers: toObject(headers),
  })
}

export default xsrf
