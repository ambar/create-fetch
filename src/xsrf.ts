import toHeaders from './utils/toHeaders'
import toObject from './utils/toObject'
import {FetchEnhancer} from './types'

const getCookie = (key: string) => {
  const [, value] = new RegExp(`${key}=([^;]+)`).exec(document.cookie) || []
  return value
}

const xsrfMethods = ['DELETE', 'PATCH', 'POST', 'PUT']

const xsrf =
  ({cookieName = '_xsrf', headerName = 'x-xsrftoken'} = {}): FetchEnhancer =>
  (fetch) =>
  (url, options = {}) => {
    const headers = toHeaders(options.headers)
    if (
      typeof document !== 'undefined' &&
      options.method &&
      xsrfMethods.indexOf(options.method.toUpperCase()) !== -1
    ) {
      const xsrfToken = getCookie(cookieName)
      if (xsrfToken) {
        headers.set(headerName, xsrfToken)
      }
    }

    return fetch(url, {
      ...options,
      headers: toObject(headers),
    })
  }

export default xsrf
