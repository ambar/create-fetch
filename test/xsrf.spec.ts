import xsrf from '../src/xsrf'

const sampleCookie = '_xsrf=uuid|foo.bar;_ga=1'

describe('xsrf', () => {
  const fetch = async (...args) => args

  beforeEach(() => {
    document.cookie = '_xsrf='
  })

  it('should add xsrf header when using POST/PATCH/POST/PUT', async () => {
    document.cookie = sampleCookie
    const myFetch = xsrf()(fetch)
    const [, {headers}] = await myFetch('/', {method: 'POST'})
    expect(headers).toMatchSnapshot()

    const methods = ['DELETE', 'PATCH', 'POST', 'PUT']
    for (const method of methods) {
      const [, options] = await myFetch('/', {method})
      expect(options.headers).toMatchObject(headers)

      const [, options2] = await myFetch('/', {method: method.toLowerCase()})
      expect(options2.headers).toMatchObject(headers)
    }
  })

  it('should not add xsrf header when using GET/HEAD/OPTIONS', async () => {
    document.cookie = sampleCookie
    const myFetch = xsrf()(fetch)
    const [, {headers}] = await myFetch('/')
    expect(headers).toMatchSnapshot()

    const methods = ['GET', 'HEAD', 'OPTIONS']
    for (const method of methods) {
      const [, options] = await myFetch('/', {method})
      expect(options.headers).toMatchObject(headers)
    }
  })

  it('should add xsrf header with custom header name', async () => {
    document.cookie = sampleCookie
    const myFetch = xsrf({headerName: 'x-xsrf-token'})(fetch)
    expect(await myFetch('/', {method: 'POST'})).toMatchSnapshot()
  })

  it('should add xsrf header with custom cookie name', async () => {
    document.cookie = 'csrf=uuid'
    const myFetch = xsrf({cookieName: 'csrf'})(fetch)
    expect(await myFetch('/', {method: 'POST'})).toMatchSnapshot()
  })

  it('should not match cookie', async () => {
    document.cookie = 'xsrf=foo;__xsrf_=bar;xsrf_=bar'
    const myFetch = xsrf()(fetch)
    expect(await myFetch('/', {method: 'POST'})).toMatchSnapshot()
  })

  it('should work without options', async () => {
    const myFetch = xsrf()(fetch)
    expect(await myFetch('/')).toMatchSnapshot()
  })
})
