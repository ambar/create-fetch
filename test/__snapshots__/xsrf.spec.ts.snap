// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`xsrf should add xsrf header when using POST/PATCH/POST/PUT 1`] = `
{
  "x-xsrftoken": "uuid|foo.bar",
}
`;

exports[`xsrf should add xsrf header with custom cookie name 1`] = `
[
  "/",
  {
    "headers": {
      "x-xsrftoken": "uuid",
    },
    "method": "POST",
  },
]
`;

exports[`xsrf should add xsrf header with custom header name 1`] = `
[
  "/",
  {
    "headers": {
      "x-xsrf-token": "uuid|foo.bar",
    },
    "method": "POST",
  },
]
`;

exports[`xsrf should not add xsrf header when using GET/HEAD/OPTIONS 1`] = `{}`;

exports[`xsrf should not match cookie 1`] = `
[
  "/",
  {
    "headers": {},
    "method": "POST",
  },
]
`;

exports[`xsrf should work without options 1`] = `
[
  "/",
  {
    "headers": {},
  },
]
`;
